src/server/stream_family.cc

```diff
@@ -843,6 +879,296 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+  } else if (cg->entries_read != SCG_INVALID_ENTRIES_READ &&
+             !streamRangeHasTombstones(s, &cg->last_id, NULL)) {
+    /* No fragmentation ahead means that the group's logical reads counter
+     * is valid for performing the lag calculation. */
+    lag = (long long)s->entries_added - cg->entries_read;
+    valid = 1;
+  } else {
+    /* Attempt to retrieve the group's last ID logical read counter. */
+    long long entries_read = streamEstimateDistanceFromFirstEverEntry(s, &cg->last_id);
+    if (entries_read != SCG_INVALID_ENTRIES_READ) {
+      /* A valid counter was obtained. */
+      lag = (long long)s->entries_added - entries_read;
+      valid = 1;
+    }
+  }
+
+  if (valid) {
+    ginfo->lag = lag;
+  }
+}
+
+void getGroupPEL(stream* s, streamCG* cg, GroupInfo* ginfo, long long count) {
+  vector<NACKInfo> nack_info_vec;
+  long long arraylen_cg_pel = 0;
+  raxIterator ri_cg_pel;
+  raxStart(&ri_cg_pel, cg->pel);
+  raxSeek(&ri_cg_pel, "^", NULL, 0);
+  while (raxNext(&ri_cg_pel) && (!count || arraylen_cg_pel < count)) {
+    streamNACK* nack = static_cast<streamNACK*>(ri_cg_pel.data);
+    NACKInfo nack_info;
+
+    streamID id;
+    streamDecodeID(ri_cg_pel.key, &id);
+    nack_info.pel_id = id;
+    nack_info.consumer_name = nack->consumer->name;
+    nack_info.delivery_time = nack->delivery_time;
+    nack_info.delivery_count = nack->delivery_count;
+
+    nack_info_vec.push_back(nack_info);
+    arraylen_cg_pel++;
+  }
+  raxStop(&ri_cg_pel);
+  ginfo->stream_nack_vec = nack_info_vec;
+}
+
+void getConsumers(stream* s, streamCG* cg, GroupInfo* ginfo, long long count) {
+  vector<ConsumerInfo> consumer_info_vec;
+  raxIterator ri_consumers;
+  raxStart(&ri_consumers, cg->consumers);
+  raxSeek(&ri_consumers, "^", NULL, 0);
+  while (raxNext(&ri_consumers)) {
+    ConsumerInfo consumer_info;
+    streamConsumer* consumer = static_cast<streamConsumer*>(ri_consumers.data);
+
+    consumer_info.name = consumer->name;
+    consumer_info.seen_time = consumer->seen_time;
+    consumer_info.pel_count = raxSize(consumer->pel);
+
+    /* Consumer PEL */
+    long long arraylen_cpel = 0;
+    raxIterator ri_cpel;
+    raxStart(&ri_cpel, consumer->pel);
+    raxSeek(&ri_cpel, "^", NULL, 0);
+    vector<NACKInfo> consumer_pel_vec;
+    while (raxNext(&ri_cpel) && (!count || arraylen_cpel < count)) {
+      NACKInfo nack_info;
+      streamNACK* nack = static_cast<streamNACK*>(ri_cpel.data);
+
+      streamID id;
+      streamDecodeID(ri_cpel.key, &id);
+      nack_info.pel_id = id;
+      nack_info.delivery_time = nack->delivery_time;
+      nack_info.delivery_count = nack->delivery_count;
+
+      consumer_pel_vec.push_back(nack_info);
+      arraylen_cpel++;
+    }
+    consumer_info_vec.push_back(consumer_info);
+    raxStop(&ri_cpel);
+  }
+  raxStop(&ri_consumers);
+  ginfo->consumer_info_vec = consumer_info_vec;
+}
+
+OpResult<StreamInfo> OpStreams(const DbContext& db_cntx, string_view key, EngineShard* shard,
```

Comment:
This function is called from the main xinfo command function. 
This function builds the StreamInfo object which contains all the info that needs to be output.
I have kept the building logic and output print logic separate.